File: akka-sample-cqrs-scala/src/main/scala/sample/cqrs/CassandraSessionExtension.scala
package sample.cqrs

import scala.concurrent.Future

import akka.Done
import akka.actor.typed.ActorSystem
import akka.actor.typed.Extension
import akka.actor.typed.ExtensionId
import akka.actor.typed.scaladsl.adapter._
import akka.cassandra.session.DefaultSessionProvider
import akka.cassandra.session.scaladsl.CassandraSession
import akka.event.Logging

object CassandraSessionExtension extends ExtensionId[CassandraSessionExtension] {

  def get(system: ActorSystem[_]): CassandraSessionExtension = apply(system)

  override def createExtension(system: ActorSystem[_]): CassandraSessionExtension =
    new CassandraSessionExtension(system)

}

class CassandraSessionExtension(system: ActorSystem[_]) extends Extension {

  val session: CassandraSession = {
    val sessionConfig = system.settings.config.getConfig("cassandra-journal")
    new CassandraSession(
      system.toClassic,
      new DefaultSessionProvider(system.toClassic, sessionConfig),
      system.executionContext,
      Logging(system.toClassic, getClass),
      metricsCategory = "sample",
      init = _ => Future.successful(Done))
  }

}

File: akka-sample-cqrs-java/src/main/java/sample/cqrs/CassandraSessionExtension.java
package sample.cqrs;

import akka.Done;
import akka.actor.typed.ActorSystem;
import akka.actor.typed.Extension;
import akka.actor.typed.ExtensionId;
import akka.actor.typed.javadsl.Adapter;
import akka.event.Logging;
import akka.persistence.cassandra.ConfigSessionProvider;
import akka.persistence.cassandra.session.CassandraSessionSettings;
import akka.persistence.cassandra.session.javadsl.CassandraSession;
import com.typesafe.config.Config;

import java.util.concurrent.CompletableFuture;

public class CassandraSessionExtension implements Extension {

  public static class Id extends ExtensionId<CassandraSessionExtension> {

    private static final Id instance = new Id();

    private Id() {}

    // called once per ActorSystem
    @Override
    public CassandraSessionExtension createExtension(ActorSystem<?> system) {
      return new CassandraSessionExtension(system);
    }

    public static CassandraSessionExtension get(ActorSystem<?> system) {
      return instance.apply(system);
    }
  }

  public final CassandraSession session;

  private CassandraSessionExtension(ActorSystem<?> system) {
    Config sessionConfig = system.settings().config().getConfig("cassandra-journal");
    this.session = new CassandraSession(
      Adapter.toClassic(system),
      new ConfigSessionProvider(Adapter.toClassic(system), sessionConfig),
      new CassandraSessionSettings(sessionConfig),
      system.executionContext(),
      Logging.getLogger(Adapter.toClassic(system), getClass()),
      "sample",
      s -> CompletableFuture.completedFuture(Done.getInstance()));
  }

}
